File: data-structures/python/src/queue.py
"""Circular queue implementation.

Note: This module is named 'queue' which shadows the stdlib queue module.
Import with care or rename to 'circular_queue' if stdlib queue is needed.
"""


class Queue:
    def __init__(self):
        self._data = [None] * 4
        self._head = 0
        self._tail = 0
        self._size = 0
        self._capacity = 4

    def enqueue(self, value):
        if self._size == self._capacity:
            self._grow()
        self._data[self._tail] = value
        self._tail = (self._tail + 1) % self._capacity
        self._size += 1

    def dequeue(self):
        if self._size == 0:
            raise IndexError("dequeue from empty queue")
        value = self._data[self._head]
        self._data[self._head] = None
        self._head = (self._head + 1) % self._capacity
        self._size -= 1
        return value

    def front(self):
        if self._size == 0:
            raise IndexError("front from empty queue")
        return self._data[self._head]

    def back(self):
        if self._size == 0:
            raise IndexError("back from empty queue")
        back_index = (self._tail - 1) % self._capacity
        return self._data[back_index]

    def size(self):
        return self._size

    def is_empty(self):
        return self._size == 0

    def clear(self):
        self._data = [None] * self._capacity
        self._head = 0
        self._tail = 0
        self._size = 0

    def copy(self):
        """Return a shallow copy of the queue."""
        clone = Queue()
        clone._capacity = self._capacity
        clone._data = [None] * self._capacity
        clone._head = self._head
        clone._tail = self._tail
        clone._size = self._size
        for i in range(self._size):
            idx = (self._head + i) % self._capacity
            clone._data[idx] = self._data[idx]
        return clone

    def _grow(self):
        new_capacity = self._capacity * 2
        new_data = [None] * new_capacity
        for i in range(self._size):
            new_data[i] = self._data[(self._head + i) % self._capacity]
        self._data = new_data
        self._head = 0
        self._tail = self._size
        self._capacity = new_capacity

    def __len__(self):
        return self._size

    def __bool__(self):
        return self._size > 0

File: data-structures/go/src/queue/queue.go
package queue

import "errors"

var ErrEmptyQueue = errors.New("Queue: empty queue")

type Queue[T any] struct {
	data     []T
	head     int
	tail     int
	size     int
	capacity int
}

func New[T any]() *Queue[T] {
	return &Queue[T]{}
}

func (q *Queue[T]) Enqueue(value T) {
	if q.size == q.capacity {
		q.grow()
	}
	q.data[q.tail] = value
	q.tail = (q.tail + 1) % q.capacity
	q.size++
}

func (q *Queue[T]) Dequeue() (T, error) {
	var zero T
	if q.size == 0 {
		return zero, ErrEmptyQueue
	}
	value := q.data[q.head]
	q.data[q.head] = zero
	q.head = (q.head + 1) % q.capacity
	q.size--
	return value, nil
}

func (q *Queue[T]) Front() (T, error) {
	var zero T
	if q.size == 0 {
		return zero, ErrEmptyQueue
	}
	return q.data[q.head], nil
}

func (q *Queue[T]) Back() (T, error) {
	var zero T
	if q.size == 0 {
		return zero, ErrEmptyQueue
	}
	idx := (q.tail - 1 + q.capacity) % q.capacity
	return q.data[idx], nil
}

func (q *Queue[T]) Size() int {
	return q.size
}

func (q *Queue[T]) IsEmpty() bool {
	return q.size == 0
}

func (q *Queue[T]) Clear() {
	var zero T
	for i := 0; i < q.capacity; i++ {
		q.data[i] = zero
	}
	q.head = 0
	q.tail = 0
	q.size = 0
}

func (q *Queue[T]) Clone() *Queue[T] {
	clone := &Queue[T]{
		data:     make([]T, q.capacity),
		head:     0,
		tail:     q.size,
		size:     q.size,
		capacity: q.capacity,
	}
	for i := 0; i < q.size; i++ {
		clone.data[i] = q.data[(q.head+i)%q.capacity]
	}
	return clone
}

func (q *Queue[T]) grow() {
	newCap := 1
	if q.capacity > 0 {
		newCap = q.capacity * 2
	}
	newData := make([]T, newCap)
	for i := 0; i < q.size; i++ {
		newData[i] = q.data[(q.head+i)%q.capacity]
	}
	q.data = newData
	q.head = 0
	q.tail = q.size
	q.capacity = newCap
}
